src/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,433 @@
+{
+    var bob = new Employee { Name = "Bob" };
+    var angela = new Employee { Name = "Angela" };
+
+    angela.Manager = bob;
+    bob.Subordinates = new List<Employee>{ angela };
+
+    var settings = new JsonSerializerSettings
+    {
+        PreserveReferencesHandling = PreserveReferencesHandling.All
+        Formatting = Formatting.Indented 
+    };
+
+    string json = JsonConvert.SerializeObject(angela, settings);
+    Console.Write(json);
+}
+```
+
+Output:
+```jsonc
+{
+    "$id": "1",
+    "Name": "Angela",
+    "Manager": {
+        "$id": "2",
+        "Name": "Bob",
+        "Subordinates": { //Note how the Subordinates' square braces are replaced with curly braces in order to include $id and $values properties, $values will now hold whatever value was meant for the Subordinates list.
+            "$id": "3",
+            "$values": [
+                {  //Note how this object denotes reference to Angela that was previously serialized.
+                    "$ref": "1"
+                }
+            ]
+        }            
+    }
+}
+```
+
+## Using Preserve on Deserialize
+On System.Text.Json:
+```cs
+public static void ReadJsonWithPreservedReferences(){
+    string json = 
+    @"{
+        ""$id"": ""1"",
+        ""Name"": ""Angela"",
+        ""Manager"": {
+            ""$id"": ""2"",
+            ""Name"": ""Bob"",
+            ""Subordinates"": {
+                ""$id"": ""3"",
+                ""$values"": [
+                    { 
+                        ""$ref"": ""1"" 
+                    }
+                ]
+            }            
+        }
+    }";
+
+    var options = new JsonSerializerOptions
+    {
+        ReferenceHandling = ReferenceHandling.Preserve
+    };
+
+    Employee angela = JsonSerializer.Deserialize<Employee>(json, options);
+    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
+}
+```
+
+On Newtonsoft's Json.Net:
+```cs
+public static void ReadJsonWithPreservedReferences(){
+    string json = 
+    @"{
+        ""$id"": ""1"",
+        ""Name"": ""Angela"",
+        ""Manager"": {
+            ""$id"": ""2"",
+            ""Name"": ""Bob"",
+            ""Subordinates"": {
+                ""$id"": ""3"",
+                ""$values"": [
+                    { 
+                        ""$ref"": ""1"" 
+                    }
+                ]
+            }            
+        }
+    }";
+
+    var options = new JsonSerializerSettings
+    {
+        MetadataPropertyHanding = MetadataPropertyHandling.Default //Json.Net reads metadata by default, just setting the option for ilustrative purposes.
+    };
+
+    Employee angela = JsonConvert.DeserializeObject<Employee>(json, settings);
+    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
+}
+```
+
+# Ground rules
+## Unsupported types
+Basically, any type that uses a `EnumerableConverter` and tries to be preserved will throw.
```

Comment:
`EnumerableConverter` is an internal type. Is it necessary to mention this? I would re-state this from the end-user's perspective and also explain why we'd throw here.

Consider moving this down below since it isn't the most important ground rule.